### DISCUSSION TITLE
How to apply constraint on the basis of count of properties?

### DISCUSSION THREAD

**Discussion by girish-tharwani** [json-schema-org/community#586] [Q&A] [answered]
Hello experts,

I am trying to write a schema that will validate json of this type

```
{
    "Steps": {
        "0": {
            "something": "something"
        },
        "1": {
            "something": "something"
        }
    }
}
```

with constraints:
1: Steps can be empty ie {"Steps":{}} is ok
2: Key values can be 0 to 99.
3: if Steps is not empty, it must have key 0 in it.

I have used patternProperties to apply the constraint on key names (constraint 2) 
```
{
    "Steps": {
        "patternProperties": {
            "^[0-9]$|^[1-9][0-9]$": {
                "type": "object",
                "$ref": "#/$defs/StepDetails"
            }
        },
        "additionalProperties": false
    }
}
```

But I dont know how to apply the constraint number 3 above. I don't want to make property "0" required unconditionally because that will interfere with my constraint number 1 so how can I specify that **"if count of properties in Steps >0, property "0" is required"** 

Thanks,
GT

**Comment by girish-tharwani** [comment]
This seem to work but I was hoping for a more elegant If-then solution

```
{
    "Steps": {
        "patternProperties": {
            "^[0-9]$|^[1-9][0-9]$": {
                "type": "object",
                "$ref": "#/$defs/StepDetails"
            }
        },
        "additionalProperties": false,
        "oneOf": [
            {
                "required": [
                    "0"
                ]
            },
            {
                "maxProperties": 0
            }
        ]
    }
}
```

**Reply by gregsdennis** [reply]
I think this is a nice solution, actually.  It's easy to understand what it's doing.

One thing you will need to do is wrap the "Steps" in a `properties` keyword.

```json
{
  "type": "object",
  "properties": {
    "Steps": { ... }
  }
}
```

**Reply by girish-tharwani** [reply]
Yes, Steps is defined like that in the schema. This was just a snippet so some portions were left out....

**Comment by jdesrosiers** [comment]
You can use `minProperties` to do a conditional on the number of properties present.

```json
"if": { "minProperties": 1 },
"then": { "required": ["0"] }
```

**Reply by girish-tharwani** [reply]
tried this. This does not work. I think "minProperties" defines the constraint of minimum number of properties required. It does not maintain the count of properties present.

**Reply by jdesrosiers** [reply, marked as answer]
It works. It expresses the constraint, "if there is at least one property (minProperties: 1),  then the '0' property is required".

If this doesn't work for you, the validator you're using probably doesn't support the `if`/`then` keywords.

```json
"Steps": {
  "type": "object",
  "patternProperties": {
    "^[0-9]$|^[1-9][0-9]$": { "type": "object" }
  },
  "additionalProperties": false,

  "if": { "minProperties": 1 },
  "then": { "required": [ "0" ] }
}
```

**Reply by girish-tharwani** [reply]
tried again and it works! Must have done something wrong last time. Thanks.
